### PR TITLE
Add BPE vocabulary

### PR DESCRIPTION
**Pull request by MorinoseiMorizo**
I added BPE vocabulary that is proposed by Sennrich et al. 2015.
Most of the codes are based on their original python implement.
But this implement can limit the vocabulary size to the specific size.

This will store all of the characters shown in training data,
so in the training data, we will not see any unknown words,
but it can appear in the test set if we use the language which has large character vocabularies like Japanese.
It may cause a problem, but currently, I haven't deal with this.

### CODE REVIEW

File: src/lib/bpe_vocabulary.cc [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+#include <boost/algorithm/string.hpp>
+#include <nmtkit/array.h>
+#include <nmtkit/corpus.h>
+#include <nmtkit/exception.h>
+
+using namespace std;
+
+namespace {
+
+struct Change {
```
Comment (odashi) [inline src/lib/bpe_vocabulary.cc]: add comment.

**Comment by MorinoseiMorizo** [comment]
fixed incorrect frequency calculation.

**Comment by MorinoseiMorizo** [comment]
fixed incorrect frequency calculation

**Comment by MorinoseiMorizo** [comment]
added test to check frequency calculation is correct.
and fixed incorrect usage of bpe_cache.
